include/zephyr/sys/ref.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_REF_H_
+#define ZEPHYR_INCLUDE_SYS_REF_H_
+
+#include <limits.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup sys_ref_apis Reference count APIs
+ * @{
+ */
+
+/* atomics are usually `long`. If they are larger, that's fine because
+ * we'll just artificially limit them . If they were smaller, we'd be unable
+ * to check if an overflow happened. This wouldn't be necessary if the atomic
+ * API told us the max value of an atomic.
+ */
+BUILD_ASSERT(sizeof(atomic_val_t) >= sizeof(long));
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when all references were released. This can also be used on SMP systems.
+ */
+struct sys_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize sys_ref at compile-time. */
+#define SYS_REF_INIT()                                                                             \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize sys_ref at runtime. */
+static inline void sys_ref_init(struct sys_ref *const ref)
+{
+	*ref = (struct sys_ref){
+		.refs = ATOMIC_INIT(1),
+	};
+}
+
+/**
+ * @brief Create a new reference to `ref`.
+ *
+ * @return The new reference.
+ */
+static inline struct sys_ref *sys_ref(struct sys_ref *const ref)
+{
+	const atomic_val_t old_rc = atomic_inc(&ref->refs);
+
+	__ASSERT_NO_MSG(old_rc <= LONG_MAX);
+
+	/* Prevent warnings with assertions disabled. */
+	(void)(old_rc);
+
+	return ref;
+}
+
+/**
+ * @brief Release the reference `ref`.
+ *
```

Comment:
Probably worth documenting the lack of synchronization here.  The decrement itself is atomic, but the use of the API is not.  For example this code looks idiomatic but is 100% broken if used in parallel:

```c
if (sys_unref(&myobj->ref)) {
    destroy_myobj(myobj);
}
```

Consider what happens if another context calls sys_ref() between the return from sys_unref() and the destroy call!

Users should be warned that they need to externally lock access to a single object so that it can't be referenced on the destroy path.  (This isn't anything in particular to do with this PR, it's just a footgun in the whole paradigm of reference counting).

Reply:
It would absolutely required an external race, but it's IMHO a common mistake with this kind of code.  The references are "atomic" so people get fooled into thinking they don't need to synchronize them, but they do.

Reply:
In `net_pkt_unref()` API, we do it like this
```
	do {
		ref = atomic_get(&pkt->atomic_ref);
		if (!ref) {
			return;
		}
	} while (!atomic_cas(&pkt->atomic_ref, ref, ref - 1));

	/* Free the pkt */
```